src/yz_solrq_worker.erl

```diff
@@ -171,89 +167,91 @@ cancel_drain(QPid) ->
-    IndexQ = enqueue(E, get_indexq(Index, State2)),
-    IndexQ2 = maybe_request_worker(Index, IndexQ),
-    IndexQ3 = maybe_start_timer(Index, IndexQ2),
-    NewState = update_indexq(Index, IndexQ3, State2),
-    ?PULSE_DEBUG("index.  NewState: ~p~n", [debug_state(NewState)]),
-    {noreply, NewState};
+init([Index, Partition]) ->
+    State0 = read_appenv(#state{index = Index, partition = Partition}),
+    State1 = get_helper(Index, Partition, State0),
+    {ok, State1}.
+
+handle_call({index, E}, From, State0) ->
+    ?PULSE_DEBUG("index.  State: ~p~n", [debug_state(State0)]),
+    State1 = maybe_purge(State0),
+    State2 = maybe_send_reply(From, State1),
+    State3 = enqueue(E, State2),
+    State4 = maybe_send_batch_to_helper(State3),
+    FinalState = maybe_start_timer(State4),
+    ?PULSE_DEBUG("index.  NewState: ~p~n", [debug_state(FinalState)]),
+    {noreply, FinalState};
 handle_call(status, _From, #state{} = State) ->
     {reply, internal_status(State), State};
 handle_call({set_hwm, NewHWM}, _From, #state{queue_hwm = OldHWM} = State) ->
     {reply, {ok, OldHWM}, maybe_unblock_vnodes(State#state{queue_hwm = NewHWM})};
 handle_call(get_hwm, _From, #state{queue_hwm = HWM} = State) ->
     {reply, HWM, State};
-handle_call({set_index, Index, Min, Max, DelayMS}, _From, State) ->
-    IndexQ = get_indexq(Index, State),
-    IndexQ2 = maybe_request_worker(Index,
-                                   IndexQ#indexq{batch_min = Min,
-                                                 batch_max = Max,
-                                                 delayms_max = DelayMS}),
-    OldParams = {IndexQ#indexq.batch_min,
-                 IndexQ#indexq.batch_max,
-                 IndexQ#indexq.delayms_max},
-    {reply, {ok, OldParams}, update_indexq(Index, IndexQ2, State)};
+handle_call({set_index, Min, Max, DelayMS}, _From,
+            State0) ->
+    State1= State0#state{batch_min = Min,
+                          batch_max = Max,
+                          delayms_max = DelayMS},
+    State2 = maybe_send_batch_to_helper(State1),
+    OldParams = {State2#state.batch_min,
+                 State2#state.batch_max,
+                 State2#state.delayms_max},
+    {reply, {ok, OldParams}, State2};
 handle_call({set_purge_strategy, NewPurgeStrategy},
     _From,
     #state{purge_strategy=OldPurgeStrategy} = State) ->
     {reply, {ok, OldPurgeStrategy}, State#state{purge_strategy=NewPurgeStrategy}};
 handle_call(cancel_drain, _From, State) ->
-    {noreply, NewState} = handle_cast(drain_complete, State),
+    NewState = handle_drain_complete(State),
     {reply, ok, NewState};
-handle_call(all_queue_len, _From, #state{all_queue_len=Len} = State) ->
+handle_call(all_queue_len, _From, #state{queue = Queue} = State) ->
+    Len = queue:len(Queue),
     {reply, Len, State};
 handle_call(reload_appenv, _From, State) ->
     {reply, ok, read_appenv(State)}.
-handle_cast({request_batch, Index, HPid}, State) ->
```

Comment:
This clause gets deleted, but then doesn't appear to be replaced by a new one like I'd expect. It looks like `request_batch/2` will end up causing the worker to crash since there's no `handle_cast` clause that will match.